grammar.go

```diff
@@ -282,6 +284,27 @@ func (g *generatorContext) parseGroup(slexer *structLexer) (node, error) {
 	return &group{expr: disj}, nil
 }
 
+// A token negation
+//
+// Accepts both the form !"some-literal" and !SomeNamedToken
+func (g *generatorContext) parseNegation(slexer *structLexer) (node, error) {
+	_, _ = slexer.Next() // advance the parser since we have '!' right now.
+	next, err := g.parseTermNoModifiers(slexer)
+	if err != nil {
+		return nil, err
+	}
+	// Now, verify that next is either a literal or a token
+	// literal / reference
+	switch next.(type) {
+	case *literal:
+		return &negation{next}, nil
+	case *reference:
+		return &negation{next}, nil
+	default:
+	}
+	return nil, fmt.Errorf("Negations can only handle literals or single tokens")
```

Comment:
No capitalisation in errors please, and this should have some context like `fmt.Errorf("only literals or references are negatable, not %s", next)